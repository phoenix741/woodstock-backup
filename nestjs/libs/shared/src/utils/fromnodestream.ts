import { AsyncIterableX } from 'ix/asynciterable';
import { safeRace } from 'ix/util/safeRace';
import { Readable } from 'stream';

const NON_FLOWING = 0;
const READABLE = 1;
const ENDED = 2;
const ERRORED = 3;

export class ReadableStreamAsyncIterable<T> extends AsyncIterableX<T> implements AsyncIterator<T> {
  private _stream: Readable;
  private _defaultSize?: number;
  private _state: number;
  private _error: any;
  private _rejectFns: Set<(err: any) => void>;
  private _endPromise: Promise<void> | undefined;

  constructor(stream: Readable, size?: number) {
    super();
    this._stream = stream;
    this._defaultSize = size;
    this._state = NON_FLOWING;
    this._error = null;
    this._rejectFns = new Set<(err: any) => void>();

    const onError = (err: any) => {
      this._state = ERRORED;
      this._error = err;
      for (const rejectFn of this._rejectFns) {
        rejectFn(err);
      }
    };

    const onEnd = () => {
      this._state = ENDED;
    };

    this._stream['once']('error', onError);
    this._stream['once']('end', onEnd);
  }

  [Symbol.asyncIterator](): AsyncIterator<T> {
    return this;
  }

  async next(size = this._defaultSize): Promise<IteratorResult<T>> {
    if (this._state === NON_FLOWING) {
      await safeRace([this._waitReadable(), this._waitEnd()]);
      return await this.next(size);
    }

    if (this._state === ENDED) {
      return { done: true, value: undefined } as any as IteratorResult<T>;
    }

    if (this._state === ERRORED) {
      throw this._error;
    }

    const value = this._stream['read'](size);
    if (value !== null) {
      return { done: false, value };
    } else {
      this._state = NON_FLOWING;
      return await this.next(size);
    }
  }

  private _waitReadable(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      const onReadable = () => {
        this._state = READABLE;
        this._rejectFns.delete(reject);
        resolve();
      };

      this._rejectFns.add(reject);
      this._stream['once']('readable', onReadable);
    });
  }

  private _waitEnd(): Promise<void> {
    if (!this._endPromise) {
      this._endPromise = new Promise<void>((resolve, reject) => {
        const onEnd = () => {
          this._state = ENDED;
          this._rejectFns.delete(reject);
          resolve();
        };

        this._rejectFns.add(reject);
        this._stream['once']('end', onEnd);
      });
    }
    return this._endPromise;
  }
}

/**
 * Creates a new async-iterable from a Node.js stream.
 *
 * @param {NodeJS.ReadableStream} stream The Node.js stream to convert to an async-iterable.
 * @param {number} [size] The size of the buffers for the stream.
 * @returns {(AsyncIterableX<T>)} An async-iterable containing data from the stream either in string or Buffer format.
 */
export function fromNodeStream<T>(stream: Readable, size?: number): AsyncIterableX<T> {
  return new ReadableStreamAsyncIterable<T>(stream, size);
}
