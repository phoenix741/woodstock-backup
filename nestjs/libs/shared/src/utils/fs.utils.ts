import { constants as constantsFs } from 'fs';
import { access } from 'fs/promises';
import { join, normalize, resolve } from 'path';

export async function isExists(path: string) {
  return access(path, constantsFs.F_OK)
    .then(() => true)
    .catch(() => false);
}

export async function findNearestPackageJson(): Promise<string | undefined> {
  let currentPath = process.cwd();
  while (currentPath !== '/') {
    const packageJson = resolve(currentPath, 'package.json');

    if (await isExists(packageJson)) {
      return packageJson;
    }

    currentPath = normalize(join(currentPath, '..'));
  }

  return undefined;
}
