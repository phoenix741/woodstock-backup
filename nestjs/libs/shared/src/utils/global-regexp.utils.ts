import { IMinimatch, Minimatch } from 'minimatch';

export function globStringToRegex(str: string): IMinimatch {
  return new Minimatch(str, { matchBase: true });
}
