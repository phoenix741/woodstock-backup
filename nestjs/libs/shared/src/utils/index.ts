export * from './concurrent-promise.util';
export * from './global-regexp.utils';
export * from './iterator.utils';
export * from './number.utils';
export * from './objects.utils';
export * from './path.utils';
export * from './rendering.utils';
