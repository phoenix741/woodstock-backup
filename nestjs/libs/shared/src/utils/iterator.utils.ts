import { AsyncIterableX, from } from 'ix/asynciterable';
import { filter } from 'ix/asynciterable/operators';
import { OperatorAsyncFunction, UnaryFunction } from 'ix/interfaces';

export function notUndefined<T>(): UnaryFunction<AsyncIterable<T | null | undefined>, AsyncIterableX<T>> {
  return function notUndefinedOperatorFunction(source: AsyncIterable<T | null | undefined>): AsyncIterableX<T> {
    return from(source).pipe(filter((x) => x != null) as OperatorAsyncFunction<T | null | undefined, T>);
  };
}
