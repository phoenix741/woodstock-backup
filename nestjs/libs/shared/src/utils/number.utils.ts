import * as Long from 'long';

declare global {
  interface BigInt {
    toJSON(): string;
  }
}

BigInt.prototype.toJSON = function () {
  return this.toString();
};

export function bigIntMax(...args: bigint[]): bigint {
  return args.reduce((m, e) => (e > m ? e : m));
}

export function bigIntMin(...args: bigint[]): bigint {
  return args.reduce((m, e) => (e < m ? e : m));
}

export function longMax(...args: Long[]): Long {
  return args.reduce((m, e) => (e.greaterThan(m) ? e : m));
}

export function longMin(...args: Long[]): Long {
  return args.reduce((m, e) => (e.lessThan(m) ? e : m));
}

export function bigIntToLong(n: bigint): Long {
  return Long.fromString(n.toString());
}

export function longToBigInt(n: Long): bigint {
  return BigInt(n.toString());
}
