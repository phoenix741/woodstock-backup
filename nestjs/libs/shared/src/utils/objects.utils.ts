export function compactArray<T>(obj: Array<T>): Array<T> {
  // eslint-disable-next-line @typescript-eslint/no-use-before-define
  return obj.filter((v) => !(v === undefined || v === null)).map((v) => compact(v as any));
}

export function compactObject<T extends object>(obj: T): T {
  for (const key in obj) {
    if (!(obj[key] === undefined || obj[key] === null)) {
      obj[key] = obj[key];
    } else {
      delete obj[key];
    }
  }

  return obj;
}

export function compact<T extends object>(obj: T | Array<T>): T | Array<T> {
  if (Array.isArray(obj)) {
    return compactArray(obj);
  } else if (typeof obj === 'object') {
    return compactObject(obj);
  } else {
    return obj;
  }
}

/**
 * Pick member of an object.
 *
 * @param obj Object to update
 * @param args List of member to keep
 */
export function pick<T, K extends keyof T>(obj: T, ...keys: K[]): Pick<T, K> {
  return keys
    .map((key) => [key, obj[key]])
    .reduce((obj, [key, val]) => Object.assign(obj, { [key as string]: val }), {} as Pick<T, K>);
}
