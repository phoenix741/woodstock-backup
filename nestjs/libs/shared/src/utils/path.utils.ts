import { sep } from 'path';
import * as tmp from 'tmp';
import { promisify } from 'util';

export const SEPARATOR = Buffer.from(sep);

export const tmpNameAsync = promisify((options: tmp.TmpNameOptions, cb: tmp.TmpNameCallback) =>
  tmp.tmpName(options, cb),
);

export function splitBuffer(buffer: Buffer, delimiter: Buffer = SEPARATOR): Buffer[] {
  let search = -1;
  const result: Buffer[] = [];

  while ((search = buffer.indexOf(delimiter)) > -1) {
    const part = buffer.subarray(0, search);
    if (part.length) {
      result.push(part);
    }
    buffer = buffer.subarray(search + delimiter.length, buffer.length);
  }

  if (buffer.length) {
    result.push(buffer);
  }

  return result;
}

export function mangle(buffer: Buffer | string): string {
  return encodeURIComponent(buffer as unknown as string);
}

export function unmangle(path: string): Buffer {
  return Buffer.from(decodeURIComponent(path), 'utf-8');
}
