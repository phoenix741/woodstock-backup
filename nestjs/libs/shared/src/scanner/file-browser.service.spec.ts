import { toArray } from 'ix/asynciterable';
import { globStringToRegex } from '../utils';
import { FileBrowserService } from './file-browser.service.js';

describe('FileBrowserService', () => {
  test('Search file in all directories', async () => {
    const service = new FileBrowserService();
    const value = (
      await toArray(service.getFiles(Buffer.from(__dirname))(Buffer.alloc(0), [], [globStringToRegex('*.spec.ts')]))
    ).sort((a, b) => a.path.compare(b.path));

    for (const file of value) {
      expect(file).toMatchSnapshot({ stats: expect.any(Object) }, 'value');
    }
  });
});
