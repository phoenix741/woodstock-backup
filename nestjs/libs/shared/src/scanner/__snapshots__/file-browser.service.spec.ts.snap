// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileBrowserService Search file in all directories: value 1`] = `
{
  "acl": [],
  "chunks": [],
  "path": {
    "data": [
      95,
      95,
      102,
      105,
      120,
      116,
      117,
      114,
      101,
      115,
      95,
      95,
    ],
    "type": "Buffer",
  },
  "stats": Any<Object>,
  "symlink": undefined,
  "xattr": {},
}
`;

exports[`FileBrowserService Search file in all directories: value 2`] = `
{
  "acl": [],
  "chunks": [],
  "path": {
    "data": [
      95,
      95,
      102,
      105,
      120,
      116,
      117,
      114,
      101,
      115,
      95,
      95,
      47,
      116,
      101,
      115,
      116,
      45,
      115,
      121,
      109,
      108,
      105,
      110,
      107,
    ],
    "type": "Buffer",
  },
  "stats": Any<Object>,
  "symlink": {
    "data": [
      46,
      46,
      47,
      105,
      110,
      100,
      101,
      120,
      46,
      116,
      115,
    ],
    "type": "Buffer",
  },
  "xattr": {},
}
`;

exports[`FileBrowserService Search file in all directories: value 3`] = `
{
  "acl": [],
  "chunks": [],
  "path": {
    "data": [
      95,
      95,
      115,
      110,
      97,
      112,
      115,
      104,
      111,
      116,
      115,
      95,
      95,
    ],
    "type": "Buffer",
  },
  "stats": Any<Object>,
  "symlink": undefined,
  "xattr": {},
}
`;

exports[`FileBrowserService Search file in all directories: value 4`] = `
{
  "acl": [],
  "chunks": [],
  "path": {
    "data": [
      95,
      95,
      115,
      110,
      97,
      112,
      115,
      104,
      111,
      116,
      115,
      95,
      95,
      47,
      102,
      105,
      108,
      101,
      45,
      98,
      114,
      111,
      119,
      115,
      101,
      114,
      46,
      115,
      101,
      114,
      118,
      105,
      99,
      101,
      46,
      115,
      112,
      101,
      99,
      46,
      116,
      115,
      46,
      115,
      110,
      97,
      112,
    ],
    "type": "Buffer",
  },
  "stats": Any<Object>,
  "symlink": undefined,
  "xattr": {},
}
`;

exports[`FileBrowserService Search file in all directories: value 5`] = `
{
  "acl": [],
  "chunks": [],
  "path": {
    "data": [
      95,
      95,
      115,
      110,
      97,
      112,
      115,
      104,
      111,
      116,
      115,
      95,
      95,
      47,
      102,
      105,
      108,
      101,
      45,
      114,
      101,
      97,
      100,
      101,
      114,
      46,
      115,
      101,
      114,
      118,
      105,
      99,
      101,
      46,
      115,
      112,
      101,
      99,
      46,
      116,
      115,
      46,
      115,
      110,
      97,
      112,
    ],
    "type": "Buffer",
  },
  "stats": Any<Object>,
  "symlink": undefined,
  "xattr": {},
}
`;

exports[`FileBrowserService Search file in all directories: value 6`] = `
{
  "acl": [],
  "chunks": [],
  "path": {
    "data": [
      102,
      105,
      108,
      101,
      45,
      98,
      114,
      111,
      119,
      115,
      101,
      114,
      46,
      115,
      101,
      114,
      118,
      105,
      99,
      101,
      46,
      116,
      115,
    ],
    "type": "Buffer",
  },
  "stats": Any<Object>,
  "symlink": undefined,
  "xattr": {},
}
`;

exports[`FileBrowserService Search file in all directories: value 7`] = `
{
  "acl": [],
  "chunks": [],
  "path": {
    "data": [
      102,
      105,
      108,
      101,
      45,
      114,
      101,
      97,
      100,
      101,
      114,
      46,
      115,
      101,
      114,
      118,
      105,
      99,
      101,
      46,
      116,
      115,
    ],
    "type": "Buffer",
  },
  "stats": Any<Object>,
  "symlink": undefined,
  "xattr": {},
}
`;

exports[`FileBrowserService Search file in all directories: value 8`] = `
{
  "acl": [],
  "chunks": [],
  "path": {
    "data": [
      104,
      97,
      115,
      104,
      45,
      114,
      101,
      97,
      100,
      101,
      114,
      46,
      116,
      114,
      97,
      110,
      115,
      102,
      111,
      114,
      109,
      46,
      116,
      115,
    ],
    "type": "Buffer",
  },
  "stats": Any<Object>,
  "symlink": undefined,
  "xattr": {},
}
`;

exports[`FileBrowserService Search file in all directories: value 9`] = `
{
  "acl": [],
  "chunks": [],
  "path": {
    "data": [
      105,
      110,
      100,
      101,
      120,
      46,
      116,
      115,
    ],
    "type": "Buffer",
  },
  "stats": Any<Object>,
  "symlink": undefined,
  "xattr": {},
}
`;
